File: java/var-core/src/test/java/dev/varar/core/DriftTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/drift.test.ts} + hash.test.ts vectors. */
class DriftTest {

    private static final Object NOOP_HANDLER = (Runnable) () -> {};

    private static Registry reg(boolean withStep) {
        Registry r = Registry.createRegistry();
        if (withStep) {
            r = Registry.addStep(r, "I withdraw {int}", "steps.ts", 1, NOOP_HANDLER, StepKind.STIMULUS);
        }
        return r;
    }

    private static Registry romanReg(boolean withStep) {
        Registry r = Registry.createRegistry();
        if (withStep) {
            r = Registry.addStep(r, "a decimal and a roman number", "steps.ts", 1, NOOP_HANDLER, StepKind.SENSOR);
        }
        return r;
    }

    private static Plan.ExecutionPlan planOf(String source, Registry r) {
        return Plan.plan(Parse.parse("w.md", source), r);
    }

    private static List<String> bare(List<Drift.Drifted> drifts) {
        List<String> out = new ArrayList<>();
        for (Drift.Drifted d : drifts) out.add(d.name() + "@" + d.line());
        return out;
    }

    /** An in-memory BaselineStore. */
    private static final class MemoryStore implements Drift.BaselineStore {
        String contents;

        @Override
        public String read() {
            return contents;
        }

        @Override
        public void write(String c) {
            contents = c;
        }
    }

    @Test
    void hashMatchesTheTypescriptVectors() {
        assertEquals("fnv1a:4f9f2cab", Hash.hashSource("hello"));
        assertEquals("fnv1a:1a47e90b", Hash.hashSource("abc"));
        assertEquals("fnv1a:4eace75e", Hash.hashSource("# Title\n"));
    }

    @Test
    void liveExamplesRecordsOneEntryPerExampleProducingParagraph() {
        Ast.VarDoc varDoc = Parse.parse("w.md", "I withdraw 40.");
        assertEquals(
                List.of(new Drift.BaselineExample("I withdraw 40", 1)),
                Drift.liveExamples(varDoc, planOf("I withdraw 40.", reg(true))));
    }

    @Test
    void deriveSpecBaselineCarriesTheFingerprint() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.SpecBaseline baseline = Drift.deriveSpecBaseline(source, varDoc, planOf(source, reg(true)));
        assertEquals(Hash.hashSource(source), baseline.sourceHash());
        assertEquals(List.of(new Drift.BaselineExample("I withdraw 40", 1)), baseline.examples());
    }

    @Test
    void noBaselineMeansNoDrift() {
        Ast.VarDoc varDoc = Parse.parse("w.md", "I withdraw 40.");
        assertTrue(Drift.detectDrift(null, varDoc, planOf("I withdraw 40.", reg(true)))
                .isEmpty());
    }

    @Test
    void aRenamedStepDrifts() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.SpecBaseline baseline = Drift.deriveSpecBaseline(source, varDoc, planOf(source, reg(true)));
        assertEquals(List.of("I withdraw 40@1"), bare(Drift.detectDrift(baseline, varDoc, planOf(source, reg(false)))));
    }

    @Test
    void anInPlaceTypoDrifts() {
        String before = "I withdraw 40.";
        Drift.SpecBaseline baseline =
                Drift.deriveSpecBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "I withdrraw 40.";
        Ast.VarDoc afterDoc = Parse.parse("w.md", after);
        assertEquals(List.of("I withdraw 40@1"), bare(Drift.detectDrift(baseline, afterDoc, planOf(after, reg(true)))));
    }

    @Test
    void aDeletedParagraphIsNotDrift() {
        String before = "I withdraw 40.";
        Drift.SpecBaseline baseline =
                Drift.deriveSpecBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        Ast.VarDoc afterDoc = Parse.parse("w.md", "");
        assertTrue(Drift.detectDrift(baseline, afterDoc, planOf("", reg(true))).isEmpty());
    }

    @Test
    void movingAndRewordingAStillMatchingExampleDoesNotDrift() {
        String before = "I withdraw 40.\n\nI withdraw 10.";
        Drift.SpecBaseline baseline =
                Drift.deriveSpecBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "I withdraw 11.\n\nI withdraw 40.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    @Test
    void moveRewordProseOnOldLineDoesNotFalsePositive() {
        String before = "I withdraw 40.";
        Drift.SpecBaseline baseline =
                Drift.deriveSpecBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "Just some notes.\n\nI withdraw 41.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    @Test
    void aParagraphRewrittenPastRecognitionIsNotDrift() {
        String before = "I withdraw 40.";
        Drift.SpecBaseline baseline =
                Drift.deriveSpecBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "The branch closed years ago.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    private static final String ROMAN =
            "Each row gives a decimal and a roman number:\n\n| decimal | roman |\n| ------: | :---- |\n| 3 | III |\n| 9 | IX |\n";

    @Test
    void headerBoundTableRecordsItsBindingParagraphOnce() {
        Ast.VarDoc varDoc = Parse.parse("r.md", ROMAN);
        assertEquals(
                List.of(new Drift.BaselineExample("Each row gives a decimal and a roman number:", 1)),
                Drift.liveExamples(varDoc, Plan.plan(varDoc, romanReg(true))));
    }

    @Test
    void aHeaderBoundBindingParagraphThatStopsMatchingDrifts() {
        Ast.VarDoc varDoc = Parse.parse("r.md", ROMAN);
        Drift.SpecBaseline baseline = Drift.deriveSpecBaseline(ROMAN, varDoc, Plan.plan(varDoc, romanReg(true)));
        assertEquals(
                List.of("Each row gives a decimal and a roman number:@1"),
                bare(Drift.detectDrift(baseline, varDoc, Plan.plan(varDoc, romanReg(false)))));
    }

    @Test
    void reconcileRecordsThenReportsAndPreservesOnDrift() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        MemoryStore store = new MemoryStore();
        assertTrue(Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false)
                .isEmpty());
        String beforeLock = store.contents;
        List<Drift.Drifted> drifts =
                Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), false);
        assertEquals(List.of("I withdraw 40@1"), bare(drifts));
        assertEquals(beforeLock, store.contents); // preserved while unacknowledged
    }

    @Test
    void reconcileUpdateModeAcceptsDrift() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        MemoryStore store = new MemoryStore();
        Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false);
        assertTrue(Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), true)
                .isEmpty());
        Drift.VarLock lock = Drift.parseVarLock(store.contents);
        assertEquals(List.of(), lock.specs().get("w.md").examples());
    }

    private static final String EXPECTED_LOCK = "{\n"
            + "  \"version\": 1,\n"
            + "  \"specs\": {\n"
            + "    \"library.md\": {\n"
            + "      \"sourceHash\": \"fnv1a:1a2b3c4d\",\n"
            + "      \"examples\": [\n"
            + "        {\n"
            + "          \"name\": \"I check out\",\n"
            + "          \"line\": 7\n"
            + "        }\n"
            + "      ]\n"
            + "    }\n"
            + "  }\n"
            + "}\n";

    @Test
    void stringifyMatchesTheTypescriptSerializerByteForByte() {
        Drift.VarLock lock = new Drift.VarLock(
                1,
                Map.of(
                        "library.md",
                        new Drift.SpecBaseline(
                                "fnv1a:1a2b3c4d", List.of(new Drift.BaselineExample("I check out", 7)))));
        assertEquals(EXPECTED_LOCK, Drift.stringifyVarLock(lock));
    }

    @Test
    void parseRoundTripsAValidLock() {
        Drift.VarLock lock = new Drift.VarLock(
                1,
                Map.of(
                        "library.md",
                        new Drift.SpecBaseline(
                                "fnv1a:1a2b3c4d", List.of(new Drift.BaselineExample("I check out", 7)))));
        Drift.VarLock parsed = Drift.parseVarLock(Drift.stringifyVarLock(lock));
        assertEquals("fnv1a:1a2b3c4d", parsed.specs().get("library.md").sourceHash());
        assertEquals(
                List.of(new Drift.BaselineExample("I check out", 7)),
                parsed.specs().get("library.md").examples());
    }

    @Test
    void parseRejectsMalformedInput() {
        assertNull(Drift.parseVarLock("not json"));
        assertNull(Drift.parseVarLock("{}"));
        assertNull(Drift.parseVarLock("{\"version\":2,\"specs\":{}}"));
        assertNull(Drift.parseVarLock("{\"version\":1,\"specs\":{\"a.md\":{\"examples\":[]}}}"));
    }

    @Test
    void driftMessageNamesTheParagraph() {
        Drift.Drifted d = new Drift.Drifted("I withdraw 40", 1, Span.spanFromOffsets("I withdraw 40.", 0, 13));
        assertTrue(Drift.message(d).contains("I withdraw 40"));
        assertFalse(Drift.message(d).isBlank());
    }
}

File: go/core/drift_test.go
package core

import (
	"fmt"
	"reflect"
	"strings"
	"testing"
)

// Port of drift.test.ts / DriftTest.java (unit-gated; drift has no golden).

func stimulusKind() *StepKind { k := Stimulus; return &k }
func sensorKind() *StepKind   { k := Sensor; return &k }

func reg(withStep bool) Registry {
	r := CreateRegistry()
	if withStep {
		next, err := AddStep(r, "I withdraw {int}", "steps.ts", 1, NoopHandler(), stimulusKind())
		if err != nil {
			panic(err)
		}
		return next
	}
	return r
}

func romanReg(withStep bool) Registry {
	r := CreateRegistry()
	if withStep {
		next, err := AddStep(r, "a decimal and a roman number", "steps.ts", 1, NoopHandler(), sensorKind())
		if err != nil {
			panic(err)
		}
		return next
	}
	return r
}

func planOf(source string, r Registry) ExecutionPlan {
	return Plan(Parse("w.md", source), r)
}

func bare(drifts []Drifted) []string {
	out := []string{}
	for _, d := range drifts {
		out = append(out, fmt.Sprintf("%s@%d", d.Name, d.Line))
	}
	return out
}

type memoryStore struct {
	contents *string
}

func (m *memoryStore) Read() (string, bool) {
	if m.contents == nil {
		return "", false
	}
	return *m.contents, true
}
func (m *memoryStore) Write(c string) { s := c; m.contents = &s }

func libraryLock() VarLock {
	return VarLock{Version: 2, Oaths: map[string]OathBaseline{
		"library.md": {SourceHash: "fnv1a:1a2b3c4d", Examples: []BaselineExample{{Name: "I check out", Line: 7}}},
	}}
}

func TestLiveExamplesRecordsOnePerParagraph(t *testing.T) {
	varDoc := Parse("w.md", "I withdraw 40.")
	got := LiveExamples(varDoc, planOf("I withdraw 40.", reg(true)))
	want := []BaselineExample{{Name: "I withdraw 40", Line: 1}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestDeriveOathBaselineCarriesFingerprint(t *testing.T) {
	source := "I withdraw 40."
	varDoc := Parse("w.md", source)
	baseline := DeriveOathBaseline(source, varDoc, planOf(source, reg(true)))
	if baseline.SourceHash != HashSource(source) {
		t.Errorf("hash mismatch")
	}
	want := []BaselineExample{{Name: "I withdraw 40", Line: 1}}
	if !reflect.DeepEqual(baseline.Examples, want) {
		t.Errorf("got %v, want %v", baseline.Examples, want)
	}
}

func TestNoBaselineMeansNoDrift(t *testing.T) {
	varDoc := Parse("w.md", "I withdraw 40.")
	if d := DetectDrift(nil, varDoc, planOf("I withdraw 40.", reg(true))); len(d) != 0 {
		t.Errorf("expected no drift, got %v", bare(d))
	}
}

func TestRenamedStepDrifts(t *testing.T) {
	source := "I withdraw 40."
	varDoc := Parse("w.md", source)
	baseline := DeriveOathBaseline(source, varDoc, planOf(source, reg(true)))
	got := bare(DetectDrift(&baseline, varDoc, planOf(source, reg(false))))
	if !reflect.DeepEqual(got, []string{"I withdraw 40@1"}) {
		t.Errorf("got %v", got)
	}
}

func TestInPlaceTypoDrifts(t *testing.T) {
	before := "I withdraw 40."
	baseline := DeriveOathBaseline(before, Parse("w.md", before), planOf(before, reg(true)))
	after := "I withdrraw 40."
	afterDoc := Parse("w.md", after)
	got := bare(DetectDrift(&baseline, afterDoc, planOf(after, reg(true))))
	if !reflect.DeepEqual(got, []string{"I withdraw 40@1"}) {
		t.Errorf("got %v", got)
	}
}

func TestDeletedParagraphIsNotDrift(t *testing.T) {
	before := "I withdraw 40."
	baseline := DeriveOathBaseline(before, Parse("w.md", before), planOf(before, reg(true)))
	afterDoc := Parse("w.md", "")
	if d := DetectDrift(&baseline, afterDoc, planOf("", reg(true))); len(d) != 0 {
		t.Errorf("expected no drift, got %v", bare(d))
	}
}

func TestMovingAndRewordingStillMatchingDoesNotDrift(t *testing.T) {
	before := "I withdraw 40.\n\nI withdraw 10."
	baseline := DeriveOathBaseline(before, Parse("w.md", before), planOf(before, reg(true)))
	after := "I withdraw 11.\n\nI withdraw 40."
	if d := DetectDrift(&baseline, Parse("w.md", after), planOf(after, reg(true))); len(d) != 0 {
		t.Errorf("expected no drift, got %v", bare(d))
	}
}

func TestMoveRewordProseOnOldLineNoFalsePositive(t *testing.T) {
	before := "I withdraw 40."
	baseline := DeriveOathBaseline(before, Parse("w.md", before), planOf(before, reg(true)))
	after := "Just some notes.\n\nI withdraw 41."
	if d := DetectDrift(&baseline, Parse("w.md", after), planOf(after, reg(true))); len(d) != 0 {
		t.Errorf("expected no drift, got %v", bare(d))
	}
}

func TestRewrittenPastRecognitionIsNotDrift(t *testing.T) {
	before := "I withdraw 40."
	baseline := DeriveOathBaseline(before, Parse("w.md", before), planOf(before, reg(true)))
	after := "The branch closed years ago."
	if d := DetectDrift(&baseline, Parse("w.md", after), planOf(after, reg(true))); len(d) != 0 {
		t.Errorf("expected no drift, got %v", bare(d))
	}
}

// depositWithdrawReg registers I withdraw, plus I deposit when withDeposit.
func depositWithdrawReg(t *testing.T, withDeposit bool) Registry {
	t.Helper()
	r := CreateRegistry()
	if withDeposit {
		next, err := AddStep(r, "I deposit {int}", "steps.ts", 1, NoopHandler(), stimulusKind())
		if err != nil {
			t.Fatal(err)
		}
		r = next
	}
	next, err := AddStep(r, "I withdraw {int}", "steps.ts", 2, NoopHandler(), stimulusKind())
	if err != nil {
		t.Fatal(err)
	}
	return next
}

// Two paragraphs that merge into one example are each a live baseline entry, and
// deleting one step def drifts only the now-prose paragraph (ADR 0012).
func TestMergedExampleParagraphsAreEachLive(t *testing.T) {
	source := "I deposit 100.\n\nI withdraw 40."
	varDoc := Parse("w.md", source)
	plan := Plan(varDoc, depositWithdrawReg(t, true))
	// One planned example (the two paragraphs merged), but two live entries.
	if len(plan.Examples) != 1 {
		t.Fatalf("expected 1 planned example, got %d", len(plan.Examples))
	}
	got := LiveExamples(varDoc, plan)
	want := []BaselineExample{{Name: "I deposit 100", Line: 1}, {Name: "I withdraw 40", Line: 3}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("live got %v, want %v", got, want)
	}
}

func TestDeletingOneStepOfMergedExampleDriftsOnlyNowProseParagraph(t *testing.T) {
	source := "I deposit 100.\n\nI withdraw 40."
	varDoc := Parse("w.md", source)
	baseline := DeriveOathBaseline(source, varDoc, Plan(varDoc, depositWithdrawReg(t, true)))
	// The deposit step is gone: its paragraph becomes prose, splitting the
	// example. The withdraw paragraph is still live; the deposit one drifts.
	got := bare(DetectDrift(&baseline, varDoc, Plan(varDoc, depositWithdrawReg(t, false))))
	if !reflect.DeepEqual(got, []string{"I deposit 100@1"}) {
		t.Errorf("got %v", got)
	}
}

const roman = "Each row gives a decimal and a roman number:\n\n| decimal | roman |\n| ------: | :---- |\n| 3 | III |\n| 9 | IX |\n"

func TestHeaderBoundRecordsBindingOnce(t *testing.T) {
	varDoc := Parse("r.md", roman)
	got := LiveExamples(varDoc, Plan(varDoc, romanReg(true)))
	want := []BaselineExample{{Name: "Each row gives a decimal and a roman number:", Line: 1}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestHeaderBoundBindingThatStopsMatchingDrifts(t *testing.T) {
	varDoc := Parse("r.md", roman)
	baseline := DeriveOathBaseline(roman, varDoc, Plan(varDoc, romanReg(true)))
	got := bare(DetectDrift(&baseline, varDoc, Plan(varDoc, romanReg(false))))
	if !reflect.DeepEqual(got, []string{"Each row gives a decimal and a roman number:@1"}) {
		t.Errorf("got %v", got)
	}
}

func TestReconcileRecordsThenReportsAndPreservesOnDrift(t *testing.T) {
	source := "I withdraw 40."
	varDoc := Parse("w.md", source)
	store := &memoryStore{}
	if d := ReconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false); len(d) != 0 {
		t.Fatalf("expected clean first run, got %v", bare(d))
	}
	beforeLock := *store.contents
	drifts := ReconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), false)
	if !reflect.DeepEqual(bare(drifts), []string{"I withdraw 40@1"}) {
		t.Errorf("got %v", bare(drifts))
	}
	if *store.contents != beforeLock {
		t.Errorf("lock should be preserved while unacknowledged")
	}
}

func TestReconcileUpdateModeAcceptsDrift(t *testing.T) {
	source := "I withdraw 40."
	varDoc := Parse("w.md", source)
	store := &memoryStore{}
	ReconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false)
	if d := ReconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), true); len(d) != 0 {
		t.Errorf("update mode should accept drift, got %v", bare(d))
	}
	lock := ParseVarLock(*store.contents)
	if lock == nil || len(lock.Oaths["w.md"].Examples) != 0 {
		t.Errorf("expected empty examples after update-mode reconcile")
	}
}

const expectedLock = "{\n  \"version\": 2,\n  \"oaths\": {\n    \"library.md\": {\n      \"sourceHash\": \"fnv1a:1a2b3c4d\",\n      \"examples\": [\n        {\n          \"name\": \"I check out\",\n          \"line\": 7\n        }\n      ]\n    }\n  }\n}\n"

func TestStringifyMatchesSerializerByteForByte(t *testing.T) {
	if got := StringifyVarLock(libraryLock()); got != expectedLock {
		t.Errorf("got:\n%s\nwant:\n%s", got, expectedLock)
	}
}

func TestParseRoundTripsValidLock(t *testing.T) {
	parsed := ParseVarLock(StringifyVarLock(libraryLock()))
	if parsed == nil {
		t.Fatal("nil parse")
	}
	oath := parsed.Oaths["library.md"]
	if oath.SourceHash != "fnv1a:1a2b3c4d" {
		t.Errorf("hash %q", oath.SourceHash)
	}
	if !reflect.DeepEqual(oath.Examples, []BaselineExample{{Name: "I check out", Line: 7}}) {
		t.Errorf("examples %v", oath.Examples)
	}
}

func TestParseRejectsMalformedInput(t *testing.T) {
	bad := []string{
		"not json",
		"{}",
		`{"version":1,"oaths":{}}`,
		`{"version":2,"oaths":{"a.md":{"examples":[]}}}`,
	}
	for _, s := range bad {
		if ParseVarLock(s) != nil {
			t.Errorf("expected nil for %q", s)
		}
	}
}

func TestDriftMessageNamesTheParagraph(t *testing.T) {
	d := Drifted{Name: "I withdraw 40", Line: 1, Span: spanFromOffsets("I withdraw 40.", 0, 13)}
	msg := DriftMessage(d)
	if !strings.Contains(msg, "I withdraw 40") {
		t.Errorf("message missing name: %q", msg)
	}
	if strings.TrimSpace(msg) == "" {
		t.Errorf("empty message")
	}
}
